src/app.js

```diff
@@ -63,64 +59,40 @@ export function app(app) {
       node,
       (node = emit("render", view)(state, actions))
     )
-
     locked = !locked
-
-    if (!loaded) {
-      emit("loaded", (loaded = true))
-    }
   }
 
-  function initialize(namespace, children, lastName) {
+  function iterate(namespace, children, lastName) {
     Object.keys(children || []).map(function(key) {
       var action = children[key]
       var name = lastName ? lastName + "." + key : key
 
       if (typeof action === "function") {
         namespace[key] = function(data) {
-          var result = emit("afterAction", {
-            name: name,
-            data: action(
-              state,
-              actions,
-              emit("beforeAction", {
-                name: name,
-                data: data
-              }).data
-            )
-          }).data
-
-          if (result == null) {
-          } else if (typeof result === "function") {
-            result = result(update)
-          } else if (typeof result.then === "function") {
-            result.then(update)
-          } else {
-            update(result)
-          }
+          emit("action", { name: name, data: data })
+
+          var result = emit("resolve", action(state, actions, data))
+
+          return typeof result === "function" ? result(update) : update(result)
```

Comment:
Is this sufficient for all situations? Because I have a problem with this example with async/await: https://codepen.io/MatejMazur/pen/eEJrJv?editors=0010 ..but maybe I'm wrong and I just don't fully get it yet. ;-)